bft_interp/src/lib.rs

```diff
@@ -142,58 +147,96 @@ where
 
         if self.head == self.cells.len() {
             if self.tape_can_grow {
                 self.cells.push(T::default());
             } else {
-                let bad_instruction = self.program.instructions()[self.program_counter].clone();
+                let bad_instruction = self.program.localised_instructions()[self.program_counter];
                 return Err(VMError::HeadOverrun(bad_instruction));
             }
         }
 
-        Ok(())
+        Ok(self.program_counter + 1)
     }
 
     /// Perform a wrapping increment on the cell pointed at by the head
-    fn increment_cell(&mut self) {
+    fn increment_cell(&mut self) -> Result<usize, VMError> {
         self.cells[self.head].wrapping_increment();
+        Ok(self.program_counter + 1)
     }
 
     /// Perform a wrapping decrement on the cell pointed at by the head
-    fn decrement_cell(&mut self) {
+    fn decrement_cell(&mut self) -> Result<usize, VMError> {
         self.cells[self.head].wrapping_decrement();
+        Ok(self.program_counter + 1)
     }
 
     /// Read a single byte from [source] and write it to the cell at head
-    fn read_value(&mut self, source: &mut impl Read) -> Result<(), VMError> {
+    fn read_value(&mut self, source: &mut impl Read) -> Result<usize, VMError> {
         let mut buffer = [0];
         match source.read_exact(&mut buffer) {
             Ok(_) => {
                 self.cells[self.head].set_value(buffer[0]);
-                Ok(())
+                Ok(self.program_counter + 1)
             }
             Err(error) => {
-                let bad_instruction = self.program.instructions()[self.program_counter].clone();
+                let bad_instruction = self.program.localised_instructions()[self.program_counter];
                 Err(VMError::from((bad_instruction, error)))
             }
         }
     }
 
     /// Print the value at head to the target output
-    fn print_value(&self, output: &mut impl Write) -> Result<(), VMError> {
+    fn print_value(&self, output: &mut impl Write) -> Result<usize, VMError> {
         let output_buf = [self.cells[self.head].get_value()];
         match output.write_all(&output_buf) {
-            Ok(_) => Ok(()),
+            Ok(_) => Ok(self.program_counter + 1),
 
             Err(error) => {
-                let bad_instruction = self.program.instructions()[self.program_counter].clone();
+                let bad_instruction = self.program.localised_instructions()[self.program_counter];
                 Err(VMError::from((bad_instruction, error)))
             }
         }
     }
+
+    /// Get the next program instruction index based on the value of the cell under the head.
+    /// If the cell is zero, return the index of the instruction after the matching ].
+    /// If the cell is not zero, return the index of the next instruction after this one.
+    fn conditional_jump_forward(&self) -> Result<usize, VMError> {
+        match self.program.jump_map()[self.program_counter] {
```

Comment:
I think I'd just make the jump map be infallible, it's not like you'll bother looking stuff up in places which aren't jumps.